### DISCUSSION TITLE
No UtteranceEnd response received

### DISCUSSION THREAD

**Discussion by Mrashidpoor** [deepgram/community#503] [General help]
### Which Deepgram product are you using?

Deepgram SDKs

### Details

Hi, I am using the hybrid model between speech_final and UtteranceEnd to realize when a speech is finalized. While I do get the speech_final message when there is silence, I do not get the UtteranceEnd message at all no matter how long I wait. I logged every single piece of data received from the websocket. Below is my code and one of the many request IDs that this happened in.

### If you are making a request to the Deepgram API and have a request ID, please paste it below:

d661207b-98bf-4891-8b49-aeb3c7b546ae

### If possible, please attach your code or paste it into the text box.

````
const endpointingDelay = '2000'
        const utteranceDelay = '2000'

        //establish the websocket connection with deepgram wss endpoint
        const deepgram = createClient(process.env.DEEPGRAM_API_KEY);
        const live = deepgram.listen.live({
            model: "nova-2-phonecall",
            punctuate: true,
            smart_format: false,
            filler_words: false,
            numerals: true,
            keywords: keywords,
            interim_results: true,
            endpointing: endpointingDelay,
            utterance_end_ms: utteranceDelay,
            encoding: 'mulaw',
            sample_rate: 8000,
        });
````

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by nikolawhallon** [comment]
Will look into this a bit more, but quick thing to try might be to change:
```
const endpointingDelay = '2000'
const utteranceDelay = '2000'
```
these values should be integers, not strings, I think (sorry, not as familiar with the SDKs as the API).

**Reply by Mrashidpoor** [reply]
Yes I did try it but that didn't work. SDK should be the same as using the API directly in terms of the request parmeters passed to create a client. I checked all over the docs but couldn't find anything useful for this issue.

**Reply by nikolawhallon** [reply]
Can you share the full output of one request which exhibits this behavior (UtteranceEnd not triggering)?

**Reply by Mrashidpoor** [reply]
Below is an output I just got from the streaming websocket:
````
{"type":"Results","channel_index":[0,1],"duration":2.7299376,"start":3.29,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0,"words":[]}]},"metadata":{"request_id":"6cbaf160-f3e1-4b0e-8984-57ea76c8566b","model_info":{"name":"2-phonecall-nova","version":"2023-11-03.29505","arch":"nova-2"},"model_uuid":"3b01cc44-e8a1-406d-ae20-cedf59c7bc63"}}
````

**Reply by Mrashidpoor** [reply]
After the system stops recognizing any words, after the utteranceDelay passed it has to send a {type: "UtteranceEnd"} object as response but it does not.

**Reply by nikolawhallon** [reply]
That seems to be a single interim result with a duration of 2.7 seconds where no words were detected, if no words are detected then no UtteranceEnd message will be sent. Are these the full results from the request? I can dig into the request id to see what might be going on.

**Reply by Mrashidpoor** [reply]
I mean, the websocket send a data object every second I can't include them all here lol. The point is that I do not get any UtteranceEnd objects as described in the docs. So yes, please look into the request ID and see if you can find any traces. Thanks in advance

**Reply by Mrashidpoor** [reply]
Any updates Nikola!

**Reply by nikolawhallon** [reply]
Sorry for the delay, I am in the PT time zone.

I looked at request `6cbaf160-f3e1-4b0e-8984-57ea76c8566b`  and the full output I get is:
```
{"type":"Results","channel_index":[0,1],"duration":1.4999375,"start":0.0,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"477a7f09-7f2a-4724-9e01-96cd489534dd","model_info":{"name":"general","version":"2023-02-22.3","arch":"base"},"model_uuid":"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"}}
{"type":"Results","channel_index":[0,1],"duration":2.4999375,"start":0.0,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"477a7f09-7f2a-4724-9e01-96cd489534dd","model_info":{"name":"general","version":"2023-02-22.3","arch":"base"},"model_uuid":"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"}}
{"type":"Results","channel_index":[0,1],"duration":3.4999375,"start":0.0,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"477a7f09-7f2a-4724-9e01-96cd489534dd","model_info":{"name":"general","version":"2023-02-22.3","arch":"base"},"model_uuid":"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"}}
{"type":"Results","channel_index":[0,1],"duration":4.4999375,"start":0.0,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"477a7f09-7f2a-4724-9e01-96cd489534dd","model_info":{"name":"general","version":"2023-02-22.3","arch":"base"},"model_uuid":"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"}}
{"type":"Results","channel_index":[0,1],"duration":3.29,"start":0.0,"is_final":true,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"477a7f09-7f2a-4724-9e01-96cd489534dd","model_info":{"name":"general","version":"2023-02-22.3","arch":"base"},"model_uuid":"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"}}
{"type":"Results","channel_index":[0,1],"duration":2.2099376,"start":3.29,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"477a7f09-7f2a-4724-9e01-96cd489534dd","model_info":{"name":"general","version":"2023-02-22.3","arch":"base"},"model_uuid":"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"}}
{"type":"Results","channel_index":[0,1],"duration":3.2099376,"start":3.29,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"477a7f09-7f2a-4724-9e01-96cd489534dd","model_info":{"name":"general","version":"2023-02-22.3","arch":"base"},"model_uuid":"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"}}
{"type":"Results","channel_index":[0,1],"duration":3.2800002,"start":3.29,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"477a7f09-7f2a-4724-9e01-96cd489534dd","model_info":{"name":"general","version":"2023-02-22.3","arch":"base"},"model_uuid":"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"}}
{"type":"Results","channel_index":[0,1],"duration":0.22993755,"start":6.57,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"","confidence":0.0,"words":[]}]},"metadata":{"request_id":"477a7f09-7f2a-4724-9e01-96cd489534dd","model_info":{"name":"general","version":"2023-02-22.3","arch":"base"},"model_uuid":"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"}}
{"type":"Metadata","transaction_key":"deprecated","request_id":"477a7f09-7f2a-4724-9e01-96cd489534dd","sha256":"fd419272e4a256622ae58fd5ded26295d95a85c39dba71fc07d88cec2cb93581","created":"2023-12-27T18:07:15.975Z","duration":6.7999377,"channels":1,"models":["96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"],"model_info":{"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4":{"name":"general","version":"2023-02-22.3","arch":"base"}}}
```
It appears no words were spoken, so no utterance began, so no utterance would have ended. Is there an example where you see explicitly a transcribed word, then 2000ms of silence after it, and no corresponding `UtteranceEnd` message?

**Reply by Mrashidpoor** [reply]
Hi Nikola no it's all good! Below is a recent request ID that I just made for testing.

bea6812c-6cad-4001-85d6-f4fc0ef99c43

**Reply by nikolawhallon** [reply]
That request seems to be firing the `UtteranceEnd` messages correctly, one example:
```
{"type":"Results","channel_index":[0,1],"duration":4.16,"start":3.02,"is_final":true,"speech_final":true,"channel":{"alternatives":[{"transcript":"this is a test for","confidence":0.9692383,"words":[{"word":"this","start":5.3743267,"end":5.61375,"confidence":0.97314453},{"word":"is","start":5.61375,"end":5.773365,"confidence":0.9614258},{"word":"a","start":5.773365,"end":6.0127883,"confidence":0.9741211},{"word":"test","start":6.0127883,"end":6.4916344,"confidence":0.96728516},{"word":"for","start":6.4916344,"end":6.9916344,"confidence":0.9692383}]}]},"metadata":{"request_id":"26cf77df-0e1b-418f-a560-95d9b7ad1638","model_info":{"name":"general","version":"2023-02-22.3","arch":"base"},"model_uuid":"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"}}
{"type":"Results","channel_index":[0,1],"duration":1.3199372,"start":7.18,"is_final":false,"speech_final":false,"channel":{"alternatives":[{"transcript":"and","confidence":0.74365234,"words":[{"word":"and","start":8.066176,"end":8.499937,"confidence":0.74365234}]}]},"metadata":{"request_id":"26cf77df-0e1b-418f-a560-95d9b7ad1638","model_info":{"name":"general","version":"2023-02-22.3","arch":"base"},"model_uuid":"96a295ec-6336-43d5-b1cb-1e48b5e6d9a4"}}
{"type":"UtteranceEnd"}
```
"for" was spoken from 6.49-6.99, then "and" was spoen from 8.06-8.49, the gap between those words is 2.07 seconds, and I correctly see the subsequent `{"type":"UtteranceEnd"}` message

**Reply by Mrashidpoor** [reply]
I guess I figured what the problem is. Because I am using the SDK I do not get all types of data when listening to the websocket. I am using the line below to listen to the events:
````
live.on(LiveTranscriptionEvents.Transcript, (data) => {}
````
Thus, it only returns the data related to Transcript events and it filters out all the other types of data. Since there are no other options available in the SDK, I guess the only solution would be to get connected to the api directly using a webocket. You might want to include this in your doumentation as well.

**Reply by diegoasua** [reply]
Experiencing the same with `live.on(LiveTranscriptionEvents.Transcript, (data) => {...}` in JS SDK 3.0.1, can't see `{"type":"UtteranceEnd"}`. Logging with metadata also does not show the utterance end message.
```javascript
 liveTranscription.on(LiveTranscriptionEvents.Metadata, async (metadata) => {
    console.log(metadata);
    });
```
What's the official (correct) way to accomplish this?

**Reply by nikolawhallon** [reply]
Ah, this makes sense! This feature isn't documented for the SDKs, I can forward this info to the SDK team for inclusion. Thanks for finding this and sorry it has been such trouble.

**Reply by diegoasua** [reply]
Patched it in a [fork](https://github.com/diegoasua/deepgram-js-sdk)

@Mrashidpoor until Deepgram fixes their SDK you can use this fork with
```js
liveTranscription.on(LiveTranscriptionEvents.UtteranceEnd, async () => {
     ...
});
```

**Reply by Mrashidpoor** [reply]
Thanks Diego but I am connecting to the websockets endpoint directly now and I think there are other features that have not been included in the SDK. Also, I found the latency to be very high for my use case and I might have to switch back to Google with better config.

**Reply by nikolawhallon** [reply]
Can you explain the use-case a bit? Deepgram streaming latencies tend to be competitive in the O(100ms) range (eg ~300ms). (There are special considerations wrt detecting end of speech which are separate issues from latency, I wrote this to explain the main strategies DG can offer https://github.com/nikolawhallon/temp-utterance-end).

**Reply by Nikhilesh-B** [reply]
I think the same issue applies for the python-SDK too @nikolawhallon, would be great if this could be integrated directly into the sdk.

**Reply by Mrashidpoor** [reply]
> Can you explain the use-case a bit? Deepgram streaming latencies tend to be competitive in the O(100ms) range (eg ~300ms). (There are special considerations wrt detecting end of speech which are separate issues from latency, I wrote this to explain the main strategies DG can offer https://github.com/nikolawhallon/temp-utterance-end).

Well, thanks for the follow up Nikola. But my problem is not with the latency specifically. My problem is with the way websocket is designed to send data every second. I send data to the model through another faster stream and it is unable to detect many words here and there, especially at the beginning of speech after a few seconds of silence. While I tried many different configs, I believe this is something beyond fix and it has to do with the way the model is designed. So that's why I had to switch back to Google and try to control wer .

**Reply by nikolawhallon** [reply]
The rate that data is sent to Deepgram has no effect on the model as we internally buffer the audio and make chunks appropriate for the model being used - the once per second messages that we send are for interim results which are not finalized and of lower quality, final results should be used when building transcripts. If there are request ids where we are missing words I'd be happy to take a look into them and try to figure out why!

**Reply by codersday** [reply]
Has this been fixed. I am facing the same issue , my request id is 123cad0b-b0bc-4860-96c6-810a5e833c46 and I donot see an UtteranceEnd in my response. I am using Node.js SDK

**Reply by jjmaldonis** [reply]
Hi @codersday, you're issue is a little different. Our logs show you sent the query parameter `sample_rate=16000` but the audio has a sample rate of `48000`. An empty transcript is often caused by an incorrect sample rate, so you if you update the sample rate to 48000 you'll get a transcript and the UtteranceEnd is likely to be sent.
